### ISSUE TITLE
MAX-ACCESS is not honored

### ISSUE THREAD

**Issue by derrell** [markabrahams/node-net-snmp#145]
It is currently possible to alter, with a SET request to the agent, a value that is defined in the MIB with MAX-ACCESS = "read-only".

**Comment by markabrahams** [comment]
Very keen to get the PR for this (#146) merged!  See the PR for detailed discussion on the implementation.

**Comment by markabrahams** [comment]
PR is merged #146 is merged and this is released in version 3.0.0 of the npm.

This contains a breaking change for hand-crafted MIB providers, hence the major version bump in the npm.  A "maxAccess" field is now mandatory in a scalar provider definition, and mandatory in each tableColumn entry of a table provider definition.

For providers produced by a MIB in the module store (via the store.getProvidersForModule() call) this should not be a breaking change for compliant MIBs, as all MIBs should contain a "MAX-ACCESS" entry in scalar and table column definitions, which is read into the provider, and so when the provider is registered with the agent, the agent will adhere to the "MAX-ACCESS" that the MIB intended.

**Comment by derrell** [comment]
I think it may not be quite right. snmpwalk doesn't traverse the tree because it gets blocked with noAccess. I think maybe GetNextRequest should not be blocked based on maxAccess, as follows:

```
@@ -4399,11 +4404,14 @@ Agent.prototype.isAllowed = function (pduType, provider, instanceNode) {
 			return maxAccess >= MaxAccess["read-write"];
 
 		case "GetRequest":
-		case "GetNextRequest":
 		case "GetBulkRequest":
 			// GetRequests require at least read-only access
 			return maxAccess >= MaxAccess["read-only"];
 
+		case "GetNextRequest":
+			// Retrieving the next OID is always allowed. Used for smpwalk.
+			return true;
+
 		default:
 			// Disallow other pdu types (TODO: verify no others needed)
 			return false;
```

**Comment by markabrahams** [comment]
I can do a snmpwalk across the entire (albeit small!) MIB in examples/snmp-agent.js.

Starting the agent like:
mark@hope:~/node-net-snmp$ node example/snmp-agent.js -p 1161

and walking like:
mark@hope:~/node-net-snmp$ snmpwalk -v 2c -c public localhost:1161 1.3.6.1
SNMPv2-MIB::sysDescr.0 = STRING: Rage inside the machine!
IF-MIB::ifIndex.1 = INTEGER: 1
IF-MIB::ifIndex.2 = INTEGER: 2
IF-MIB::ifDescr.1 = STRING: lo
IF-MIB::ifDescr.2 = STRING: eth0
IF-MIB::ifType.1 = INTEGER: softwareLoopback(24)
IF-MIB::ifType.2 = INTEGER: myrinet(99)
IF-MIB::ifType.2 = No more variables left in this MIB View (It is past the end of the MIB tree)

succeeds.

Can you post a sample provider definition or MIB state where an snmpwalk fails?

**Comment by derrell** [comment]
Yes, that example works for me too. I will try to create a simple example, tomorrow, using my MIB, that causes the problem.

**Comment by markabrahams** [comment]
In theory, the order of the agent's request processing should prevent this from being an issue.

The agent.getNextRequest() and agent.getBulkRequest() calls (used for snmpwalk) first calculate the OIDs for the new "next" varbinds and call agent.request().  maxAccess is only checked in agent.request(), which will check against the correct OIDs.

Obviously there's a problem that you're getting that needs attention - but just pointing out there shouldn't be a fundamental design problem here.

**Comment by markabrahams** [comment]
Hi @derrell - I've checked PR #148 and the agent is doing the correct thing (well - almost!).

This entry (OID 1.0.15628.4.1.4.1.1.1) is "not-accessible":

rsuSRMIndex OBJECT-TYPE
    SYNTAX       RsuTableIndex
    MAX-ACCESS   not-accessible
    STATUS       current
    DESCRIPTION
        "Store and Repeat Message Index"
    ::= { rsuSRMStatusEntry 1 }

and is almost correctly returned by the agent as such:

get-response
    request-id: 1584978604
    error-status: noAccess (6)
    error-index: 0
    variable-bindings: 1 item
        1.0.15628.4.1.4.1.1.1: Value (Null)
            Object Name: 1.0.15628.4.1.4.1.1.1 (iso.0.15628.4.1.4.1.1.1)
            Value (Null)

I say almost, as the only fix the agent needs is to set the error-index correctly on this error condition.  The above shows that the error-index was being set to zero, but should be set to the one-based index of the failed varbind as per RFC 3416 Section 4.2.4.

I've just made the fix to set error-index correctly now in master, and published as version 3.0.1 of the npm.

**Comment by markabrahams** [comment]
Note that the above fix doesn't prevent the same result from the Net-SNMP tools "snmpwalk" command you're using for testing.  This command stops the walk on reception of a noAccess error, and this will still occur even with the error-index fix.

Now putting a "non-accessible" index entry in a MIB table is pretty antisocial behaviour, as it breaks the common ways of reading values from the table (you've hit a break for "snmpwalk", but other tools/libraries will similarly struggle).  First, if you have any sway over the contents of the MIB, I'd feed back to change MAX-ACCESS to that entry (and any similar entries) from "non-accessible" to "read-only".

This also highlights a concern with breakages people might experience with this implementation - and hence I guess addressing it with a "breaking change" major version bump.  Library users will no longer get away with poorly designed MIBs "just working".  So a bit more pain for users at the expense of enforcing correctness/precision in MIB and provider definitions.

I'm hoping the pain is not too great, but if the author of the feature has hit it, perhaps this is an ominous sign haha! :-)

**Comment by markabrahams** [comment]
A workaround for this is to patch the provider definition after reading in from the MIB, changing maxAccess to "read-only" for such entries before registering the provider with the agent.

**Comment by derrell** [comment]
Thanks, @markabrahams! I've implemented the work-around for each of the table indexes.

I think I'm a few years too late to the party, to be suggesting changes to the MIB. Although it is in only very limited use in production, there have been a number of field trials, and various parties have implemented to this MIB and there's a conformance certification agency (of the whole spec; the MIB is only one component). I will see if it's possible to request/suggest changes to the MIB.

I am curious how/why other implementers haven't seen that issue, as most seem to be implementing using net-snmp, which presumably properly implements not-accessible. Maybe no one is walking the tree???

Anyway, I will close this issue now. Thanks again!

**Comment by derrell** [comment]
Huh... It seems that the index fields, when used as it is in the RSU MIB (e.g., rsuSRMIndex, referenced above), _must_ be "not-accessible", according to RFC-2578, near the end of section 7.7:

> Objects which are both specified in the INDEX clause of a conceptual
> row and also columnar objects of the same conceptual row are termed
> auxiliary objects.  The MAX-ACCESS clause for auxiliary objects is
> "not-accessible", except in the following circumstances:
> 
> (1)  within a MIB module originally written to conform to SMIv1, and
>      later converted to conform to SMIv2; or
> 
> (2)  a conceptual row must contain at least one columnar object which is
>      not an auxiliary object.  In the event that all of a conceptual
>      row's columnar objects are also specified in its INDEX clause, then
>      one of them must be accessible, i.e., have a MAX-ACCESS clause of
>      "read-only". (Note that this situation does not arise for a
>      conceptual row allowing create access, since such a row will have a
>      status column which will not be an auxiliary object.)
> 

So the RSU MIB seems to be correct. Does that mean that it can never be walked and that that's ok? Or does it mean that a walk should somehow be allowed?

**Comment by markabrahams** [comment]
Good spot @derrell !  That's another level of detail I failed to appreciate!

Immediately after the section you quoted is:

>    Note that objects specified in a conceptual row's INDEX clause need
>    not be columnar objects of that conceptual row.  In this situation,
>    the DESCRIPTION clause of the conceptual row must include a textual
>    explanation of how the objects which are included in the INDEX clause
>    but not columnar objects of that conceptual row, are used in uniquely
>    identifying instances of the conceptual row's columnar objects.

Which falls short of explaining what on further research seems to be the common method of handling potential auxiliary objects.  I say common, but I've only looked at "sysORTable" (has a "non-accessible" sysORIndex) from SNMPv2-MIB, how Net-SNMP handles this, and made sweeping assumptions!  (also you can walk the entire MIB of the Net-SNMP agent without the NoAccess error, which means there either there must not be a single non-accessible entry in it, or they're handling non-accessibles in a different way than returned NoAccess).

So it looks like Net-SNMP doesn't include the auxiliary object as a column in the table (which according to the wording makes them not "auxiliary objects", as the wording defines these as objects that are both in the INDEX clause *and* columnar objects of the same row).  The RFC says the INDEX clause object "need not be" a columnar object.  What it fails to say is the criteria for making a decision to include a non-accessible INDEX clause object as an actual "column".  When should it be a column, and when shouldn't it be a column?  But given that our current implementation (leaving non-accessible entries as table columns) that breaks GetNext, walks, etc. is problematic, it seems that we want to copy Net-SNMP's approach and just not add auxiliary objects as non-accessible columns in a row, instead leaving them out of the row altogether.

The intention seems discernible here - you don't double up on index information as a column, when it's already derivable by the index information in the OID.  I can't find an explicit mention/discussion of that in the RFCs, but it would make sense as motivation.

So...the "fix" here is to not add non-accessible table sequence objects as "columns" in a table when adding a row.  I'm hoping it's not too major, but there is a lot of table indexing code to cater for off-table indexes, composite indexes, non-integer indexes, etc. so we just need to be a bit careful with index-related changes like this.  But it does need to be done, as what we've got seems to be breaking behaviour that needs fixing.

**Comment by markabrahams** [comment]
And apologies to the RSU MIB people for the "antisocial" comment above - when they were actually complying with the RFC!  it seems the antisocial behaviour is coming from this library itself! :-)  And not helped by the RFC being a little light on direction here.

**Comment by derrell** [comment]
@markabrahams You're reading the RFC a bit differently than I did. You're thinking, I believe, that the agent can/should move columnar index objects out of a table to make them non-auxiliary objects. My reading, on the other hand, was that MIBs could/should be designed with the index objects residing outside of the table; not that the agent should move them from where the MIB says they live. The RFC specifically states that they can't be scalar objects but can be columns of a different table. (Sorry, I don't have the spec here to reference the section but I think it was that same 7.7) Your reading would solve my earlier-stated conundrum, that there are existing implementations of the RSU MIB that I would be surprised if they couldn't be walked. I just feel that your interpretation seems like a stretch...

**Comment by derrell** [comment]
It would be really nice to find an existing implemention of net-snmp with a MIB that defines a non-accessible index column in a table, and see how net-snmp handles it. Do you know of any such implementations?

**Comment by markabrahams** [comment]
Yep, as mentioned above, check out the Net-SNMP agent's implementation of sysORTable, which has a non-accessible sysORIndex.  You'll notice that the sysORIndex isn't there as a table column.

My thinking is not so much RFC interpretation as it is RFC interpolation.  The RFC is silent on whether INDEX clause objects *should* appear as a table column or not.

So given the lack of RFC direction, it's my penchant to align with some "defacto standard" Net-SNMP behaviour - and that is influencing the direction of my thinking here.  Net-SNMP doesn't include non-accessible INDEX clause objects in the table rows, so I'm suggesting we do the same.  And this removes the problems with GetNext, walk, etc. with our current behaviour to include non-accessible objects and return NoAccess for them.

Even the whole piece of the RFC you've quoted on auxiliary objects doesn't apply if we - like Net-SNMP - do not add the non-accessible INDEX clause object to the table row, as by the RFC definition that makes them no longer auxiliary objects!

**Comment by derrell** [comment]
Talked through, and I'm aligned. Sounds good. Thanks!

**Comment by markabrahams** [comment]
I've made the change now to prevent the addition of non-accessible index objects to table rows, and published in version 3.0.4 of the npm.

On reviewing the existing index code and doing some testing, it seems that the other index-related code is all on the OID side, which is unaffected, making the change for this quite small after all - really just a single extra test in addTableRow (although I fixed a couple of unrelated null-related errors while I was at it).

I'm a little uncomfortable with the inconsistency in having not-accessible *non-index* objects still added to the MIB, but on the other hand, I'd be a little uncomfortable with leaving them out as well.  After all, what are these defined for at all if not adding to the MIB tree?

@derrell - if you could review and close if you're happy, that would be great thanks!

**Comment by derrell** [comment]
LGTM. I've tested with RSU-MIB, without my kludge th change the not-accessible indexes to read-only, and I'm able to walk the mib. (It dies late, with a noAccess, but that seems unrelated.)

Closing. Thanks @markabrahams!
